pkg/credentialprovider/provider.go

```diff
@@ -90,8 +90,8 @@ func (d *CachingDockerConfigProvider) Provide(image string) DockerConfig {
 	d.mu.Lock()
 	defer d.mu.Unlock()
 
-	// If the cache hasn't expired, return our cache
-	if time.Now().Before(d.expiration) {
+	// If the cache hasn't expired, return our cache if cache is not empty
+	if time.Now().Before(d.expiration) && len(d.cacheDockerConfig) > 0 {
```

Comment:
this is problematic if `d.Provider.Provide` is expensive, and ends up returning an empty entry

Reply:
@liggitt what about add a new field(`useEmptyCache`) in `CachingDockerConfigProvider`,  it's `true` by default for aws, gce, and `false` for azure? and change following interface a little:
`Provide(image string) DockerConfig`  --> `Provide(image string) (DockerConfig, useEmptyCache)`

Reply:
rather than an interface change (lower layers should probably not be aware of a higher-level cache), an option in CachingDockerConfigProvider that defaults to current behavior seems ok.

```diff
diff --git a/pkg/credentialprovider/provider.go b/pkg/credentialprovider/provider.go
index 72a706b5c74..db283d63566 100644
--- a/pkg/credentialprovider/provider.go
+++ b/pkg/credentialprovider/provider.go
@@ -58,6 +58,10 @@ type CachingDockerConfigProvider struct {
        Provider DockerConfigProvider
        Lifetime time.Duration
 
+       // Cacheable is an optional function that returns true if the specific config should be cached.
+       // If nil, all configs are treated as cacheable.
+       Cacheable func(DockerConfig) bool
+
        // cache fields
        cacheDockerConfig DockerConfig
        expiration        time.Time
@@ -96,7 +100,10 @@ func (d *CachingDockerConfigProvider) Provide(image string) DockerConfig {
        }
 
        klog.V(2).Infof("Refreshing cache for provider: %v", reflect.TypeOf(d.Provider).String())
-       d.cacheDockerConfig = d.Provider.Provide(image)
-       d.expiration = time.Now().Add(d.Lifetime)
-       return d.cacheDockerConfig
+       config := d.Provider.Provide(image)
+       if d.ShouldCache == nil || d.ShouldCache(config) {
+               d.cacheDockerConfig = config
+               d.expiration = time.Now().Add(d.Lifetime)
+       }
+       return config
 }
```

then azure can set 
```go
Cacheable: func(d DockerConfig) bool { return len(d) > 0 }
```

Reply:
I just followed your instructions, PTAL, thanks!